EIPS/eip-5164.md

```diff
@@ -0,0 +1,209 @@
+
+This specification allows contracts on one chain to send messages to contracts on another chain. There are two key interfaces that needs to be implemented:
+
+- CrossChainRelayer
+- CrossChainExecutor
+
+The `CrossChainRelayer` lives on the origin chain. Users can share a single `CrossChainRelayer` or deploy their own.
+
+The `CrossChainExecutor` lives on the destination chain and executes relayed calls. Users can share a single `CrossChainExecutor` or deploy their own.
+
+### CrossChainRelayer
+
+The `CrossChainRelayer` lives on the chain from which messages are sent. The Relayer's job is to broadcast the messages through a transport layer.
+
+#### Methods
+
+**relayCalls**
+
+MUST emit the `RelayedCalls` event when successfully called.
+
+MUST increment a `nonce` so that each batch of calls can be uniquely identified.
+
+MAY require payment. Some bridges may require payment in the native currency, so the function is payable.
+
+```solidity
+struct Call {
+  address target;
+  bytes data;
+}
+
+interface CrossChainRelayer {
+    function relayCalls(CrossChainExecutor executor, Call[] calldata calls, uint256 gasLimit) external payable;
+}
+```
+
+```yaml
+- name: relayCalls
+  type: function
+  stateMutability: payable
+  inputs:
+    - name: executor
+      type: CrossChainExecutor
+    - name: calls
+      type: Call[]
+    - name: gasLimit
+      type: uint256
+```
+
+#### Events
+
+**RelayedCalls**
+
+MUST be emitted by the `CrossChainRelayer` when `relayCalls` is called.
+
+```solidity
+interface CrossChainRelayer {
+  event RelayedCalls(
+      uint256 indexed nonce,
+      address indexed sender,
+      CrossChainExecutor indexed executor,
+      Call[] calls,
+      uint256 gasLimit
+  );
+}
+```
+
+```yaml
+- name: RelayedCalls
+  type: event
+  inputs:
+    - name: nonce
+      indexed: true
+      type: uint256
+    - name: sender
+      indexed: true
+      type: address
+    - name: executor
+      indexed: true
+      type: address
+    - name: calls
+      type: Call[]
+    - name: gasLimit
+      type: uint256
+```
+
+#### Error handling
+
+**GasLimitTooHigh**
+
+MAY revert if the `gasLimit` passed to `relayCalls` is higher than the maximum gas limit accepted by the bridge being used.
+
+```solidity
+interface CrossChainRelayer {
+  error GasLimitTooHigh(
+    uint256 gasLimit,
+    uint256 maxGasLimit
+  );
+}
+```
+
+### CrossChainExecutor
+
+The `CrossChainExecutor` executes relayed calls. Developers must implement a `CrossChainExecutor` in order to execute messages on the receiving chain.
+
+#### Authentication
+
+SHOULD authenticate that the call has been performed by the bridge transport layer.
+
+SHOULD follow [EIP-2771](./eip-2771.md) and append the address of the Transaction Signer (20 bytes of data) on the origin chain to the end of the call data being executed.
```

Comment:
These are incomplete sentences. Please fix this.

Reply:
Why are these sentences incomplete? Which format am I supposed to follow?

Reply:
These are sentence fragments; they are missing a subject. I'll fix them for you.

Reply:
Ah ok thanks. I was mainly following the style of other EIPs.